File: rxjavaroom/src/main/java/com/ypw/rxjavaroom/Injection.java
package com.ypw.rxjavaroom;

import android.content.Context;
import com.ypw.rxjavaroom.persistence.LocalUserDataSource;
import com.ypw.rxjavaroom.persistence.UserDatabase;
import com.ypw.rxjavaroom.ui.ViewModelFactory;

/**
 * @author : fengzili on
 * @email : 291924028@qq.com
 * @date : 2019/10/24 0024
 * @pkn : com.ypw.rxjavaroom
 * @desc :
 */
public class Injection {
    public static UserDataSource provideUserDataSource(Context context) {
        UserDatabase database = UserDatabase.getInstance(context);
        return new LocalUserDataSource(database.userDao());
    }

    public static ViewModelFactory provideViewModelFactory(Context context) {
        UserDataSource dataSource = provideUserDataSource(context);
        return new ViewModelFactory(dataSource);
    }
}

File: paging/src/main/java/com/paging/Injection.kt
package com.paging

import android.content.Context
import androidx.lifecycle.ViewModelProvider
import com.paging.api.GithubService
import com.paging.data.GithubDataRepository
import com.paging.db.GithubLocalRepository
import com.paging.db.RepoDatabase
import com.paging.ui.ViewModelFactory
import java.util.concurrent.Executors

/**
 *  @author : fengzili on
 *  @email  : 291924028@qq.com
 *  @date   : 2019/12/16 0016
 *  @pkn    : com.paging
 *  @desc   :
 */
object Injection {

    private fun provideGithubRepository(context: Context): GithubDataRepository {
        return GithubDataRepository(GithubService.create(), provideCache(context))
    }

    private fun provideCache(context: Context): GithubLocalRepository {
        val database = RepoDatabase.getInstance(context)
        return GithubLocalRepository(database.repoDao(), Executors.newSingleThreadExecutor())
    }

    fun provideViewModelFactory(context: Context): ViewModelProvider.Factory {
        return ViewModelFactory(provideGithubRepository(context))

    }
}